### PR TITLE
Feature/112 reconfigure the detectors to use global code for the binary serializers 2

### PR DESCRIPTION
**Pull request by hayakawa16**
Creating a draft PR so it goes through all of the checks.  Will add reviewers next week.

### CODE REVIEW

**Comment by lmalenfant** [comment]
@hayakawa16 The quality gate passed!

There are 11 new code smells, we can't fix them all (some are cognitive complexity with the factory code) but some of them can be fixed.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs:437]: Remove extra line.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs:440]: Remove extra line.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfRhoAndSubregionHistDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfRhoAndSubregionHistDetector.cs:391]: Remove extra line.

Add a line before the comments.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetector.cs:413]: Remove extra line.

Add line above comments.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfRhoAndSubregionHistDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedMTOfRhoAndSubregionHistDetector.cs:281]: Remove extra line.

Add line above comments.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs:291]: Remove extra line.

Add line above comments.

File: src/Vts/MonteCarlo/Detectors/dMCdROfRhodMuaDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/dMCdROfRhodMuaDetector.cs:286]: Add line above comments.

File: src/Vts/MonteCarlo/Detectors/dMCdROfRhodMusDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/dMCdROfRhodMusDetector.cs:310]: Add line above comments.

File: src/Vts/MonteCarlo/Detectors/pMCROfFxAndTimeDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfFxAndTimeDetector.cs:243]: Add line above comments.

File: src/Vts/MonteCarlo/Detectors/pMCROfFxDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfFxDetector.cs:226]: Add a line.

File: src/Vts/MonteCarlo/Detectors/pMCROfRhoAndTimeDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfRhoAndTimeDetector.cs:226]: Add a line.

File: src/Vts/MonteCarlo/Detectors/pMCROfRhoAndTimeRecessedDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfRhoAndTimeRecessedDetector.cs:240]: Add a line.

File: src/Vts/MonteCarlo/Detectors/pMCROfXAndYAndTimeAndSubregionDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfXAndYAndTimeAndSubregionDetector.cs:288]: Remove line.

Add line above the comments.

File: src/Vts/MonteCarlo/Detectors/ROfXAndYAndTimeDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/ROfXAndYAndTimeDetector.cs:206]: Add a line above the comments.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedMTOfXAndYAndSubregionHistDetectorTests.cs
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedMTOfXAndYAndSubregionHistDetectorTests.cs:54]: Fixed

File: src/Vts.Test/Unit/MonteCarlo/Detectors/SurfaceFiberTests.cs
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/SurfaceFiberTests.cs:1]: Another good catch.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetectorTests.cs
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetectorTests.cs:45]: Fixed

File: src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs:45]: Fixed

File: src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetectorTests.cs
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetectorTests.cs:54]: Fixed

File: src/Vts.Test/Unit/MonteCarlo/Detectors/pMCROfFxAndTimeTests.cs
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/pMCROfFxAndTimeTests.cs:1]: Fixed.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/pMCROfXAndYAndTimeAndSubregionDetectorTests.cs
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/pMCROfXAndYAndTimeAndSubregionDetectorTests.cs:46]: Fixed

File: src/Vts/MonteCarlo/Detectors/ATotalDetector.cs
Comment (hayakawa16) [reply src/Vts/MonteCarlo/Detectors/ATotalDetector.cs:3]: Done.

**Review by hayakawa16 — COMMENTED** [review]
Thank you for your very detailed review!  I made code changes for all suggestions, retested and pushed.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedDynamicMTOfFxAndSubregionHistDetectorTests.cs
Comment (lmalenfant) [inline src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedDynamicMTOfFxAndSubregionHistDetectorTests.cs:111]: 2 statements on one line.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs
Comment (lmalenfant) [inline src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs:86]: Extra line.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs
Comment (lmalenfant) [inline src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs:85]: Extra line.

File: src/Vts/MonteCarlo/Detectors/ROfRhoAndOmegaDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/ROfRhoAndOmegaDetector.cs:200]: Add line above comments.

File: src/Vts/MonteCarlo/Detectors/ROfXAndYAndMaxDepthDetector.cs
Comment (lmalenfant) [reply src/Vts/MonteCarlo/Detectors/ROfXAndYAndMaxDepthDetector.cs:205]: Line is still missing here.

File: src/Vts/MonteCarlo/Detectors/ROfXAndYAndTimeAndSubregionDetector.cs
Comment (lmalenfant) [reply src/Vts/MonteCarlo/Detectors/ROfXAndYAndTimeAndSubregionDetector.cs:251]: Line is still missing.

File: src/Vts/MonteCarlo/Detectors/ReflectedDynamicMTOfFxAndSubregionHistDetector.cs
Comment (lmalenfant) [reply src/Vts/MonteCarlo/Detectors/ReflectedDynamicMTOfFxAndSubregionHistDetector.cs:415]: Having a line before the closing bracket of an object definition could result in a maintainer putting code in the wrong place.

File: src/Vts/MonteCarlo/Detectors/ReflectedMTOfXAndYAndSubregionHistDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/ReflectedMTOfXAndYAndSubregionHistDetector.cs:293]: Line before the closing bracket.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs
Comment (lmalenfant) [reply src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs:434]: Extra line before the closing bracket.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfRhoAndSubregionHistDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedMTOfRhoAndSubregionHistDetector.cs:278]: Extra line before the closing bracket.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs:289]: Extra line before the closing bracket.

**Review by lmalenfant — CHANGES_REQUESTED** [review]
I have added a few more comments where there is a line feed before the end of an object definition, it is not intentional whitespace and I think it could result in a developer putting future code in the wrong place.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs
Comment (hayakawa16) [reply src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs]: All review comments implemented.  Thank you for all of your help!

**Comment by hayakawa16** [comment]
Thanks Lisa!  I will await dcuccia's review before proceeding.

**Comment by hayakawa16** [comment]
Just wanted to post that I took an infile that I ran with master on 10/12/23 and reran it using this branch.  The results are identical!  It was a sanity check suggested by lmalenfant and gives us more confidence in our code modifications.

**Review by dcuccia — APPROVED** [review]
Looks great - thanks for all the hard work that went into this. My one comment that we should consider changing before merging is that we update the names of the test methods. 'Validate_deserialized_class_is_correct_when_using_GetBinarySerializers` is less accurate than `Validate_deserialized_binary_arrays_are_correct_when_using_GetBinarySerializers`.

My comments regarding things we might consider for the future are:

1. That we refactor the test projects into separate Integration and Unit projects (and as part of that, generally better align test names to the classes/class members they're testing).
2. That we pare down `Vts.Test.MonteCarlo.DetectorIOTests` significantly, now that all detectors' serializers' behavior/correctness are already covered by these new unit tests (don't need duplicate checks - wastes time and increases software maintenance load). Remaining work would probably be to refactor what remains to unit tests that test the detector-specific JSON serialization/deserialization for the non-array properties.
3. Similarly, to refactor test classes in the `Vts.Test.MonteCarlo.Detectors` into their Unit counterparts, if feasible - that way we can declare + test the intended behavior of the specific detector without relying on executing a Monte Carlo simulation (i.e. correct tally accumulation and normalization, e.g. focus on correctness of code, not accuracy of MC).
4. Once 1 & 2 are done, we would be left with tests that test the accuracy of the Monte Carlo for those detectors, which would be probably best labeled as integration tests. (An implementation detail, but a potential way to do that rapidly be to use an in-memory post-processor database, so you don't have to "run" an MC for each detector, just walk through the photon data points).

Perhaps we could create a couple follow-on placeholder tickets (to work on when appropriate) that documents our desire to do the above?

**Comment by hayakawa16** [comment]
dcuccia, thank you for your review!  I can easily modify the unit test names. 

I like your ideas for future work.  It always bothered me how many times a MC simulation is executed during the course of running the unit tests.  If this could be optimized somehow, that would be great!  In addition to your list, in our discussion on this issue we uncovered the need to analyze PopulateFromEnumerable and possibly update application to be column-major.  Creating place holder issues for these features sounds good.  Would you like to create them?

**Pull request merged**
